### ISSUE TITLE
Memory leaks in examples

### ISSUE THREAD

**Issue by lostman** [WebAssembly/wasm-c-api#41]
When compiled with `-fsanitize=address`:
```
==== C hello ====
Initializing...
Loading binary...
Compiling module...
Creating callback...
Instantiating module...
Extracting export...
Calling export...
Calling back...
> Hello World!
Shutting down...
Done.

=================================================================
==2955==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 24 byte(s) in 1 object(s) allocated from:
    #0 0x518610  (/stuff/dfinity/wasm-c-api/out/example/hello-c+0x518610)
    #1 0x547487  (/stuff/dfinity/wasm-c-api/out/example/hello-c+0x547487)
    #2 0x530f12  (/stuff/dfinity/wasm-c-api/out/example/hello-c+0x530f12)
    #3 0x53134f  (/stuff/dfinity/wasm-c-api/out/example/hello-c+0x53134f)

Indirect leak of 48 byte(s) in 1 object(s) allocated from:
    #0 0x518980  (/stuff/dfinity/wasm-c-api/out/example/hello-c+0x518980)
    #1 0x52c507  (/stuff/dfinity/wasm-c-api/out/example/hello-c+0x52c507)
    #2 0x548a6e  (/stuff/dfinity/wasm-c-api/out/example/hello-c+0x548a6e)

Indirect leak of 40 byte(s) in 1 object(s) allocated from:
    #0 0x518610  (/stuff/dfinity/wasm-c-api/out/example/hello-c+0x518610)
    #1 0x5312f4  (/stuff/dfinity/wasm-c-api/out/example/hello-c+0x5312f4)

SUMMARY: AddressSanitizer: 112 byte(s) leaked in 3 allocation(s).
Makefile:104: recipe for target 'run-hello-c' failed
make: *** [run-hello-c] Error 1
rm out/example/hello-c.o
```

It's not obvious where the leak is coming from or how to clean it up.

**Comment by rossberg** [comment]
That is odd, I cannot reproduce that, it's running fine on my machine (I always run with both -fsanitize=address and -DDEBUG).

Can you build with debug symbols, to get some line numbers?

**Comment by lostman** [comment]
I'll try to reproduce in a clean environment with Docker.

This should've been compiled with `-DDEBUG` (that's the default in the `Makefile`).

**Comment by hishamhm** [comment]
We ran into the same issue here. We just tried building the examples on a Mac and on Linux and the above errors happened on Linux only. Data points:

* LeakSanitizer failures observed on Linux/x86_64 with `clang -v`: `clang version 7.0.1 (tags/RELEASE_701/final) - Target: x86_64-unknown-linux-gnu` (Ubuntu 14.04 binaries from clang.llvm.org)
* no failures observed with Clang on Mac 10.14 with `clang -v`: `Apple LLVM version 10.0.0 (clang-1000.11.45.5) - Target: x86_64-apple-darwin18.0.0`

**Comment by hishamhm** [comment]
Linux output:

```
make run-hello-c

==== C hello ====
Initializing...
Loading binary...
Compiling module...
Creating callback...
Instantiating module...
Extracting export...
Calling export...
Calling back...
> Hello World!
Shutting down...
Done.

=================================================================
==18462==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 808 byte(s) in 1 object(s) allocated from:
    #0 0x5021e2 in operator new[](unsigned long, std::nothrow_t const&) /local/mnt/workspace/bcain_clang_bcain-ubuntu_14427/llvm/utils/release/final/llvm.src/projects/compiler-rt/lib/asan/asan_new_delete.cc:115:3
    #1 0x53dfab in wasm::StoreImpl::make_handle() /Users/hisham/wasm-c-api/./src/wasm-v8.cc:390:22
    #2 0x530526 in wasm_module_new /Users/hisham/wasm-c-api/src/wasm-c.cc:703:18
    #3 0x504e0a in main /Users/hisham/wasm-c-api/example/hello.c:46:31
    #4 0x7f248b94e290 in __libc_start_main /sources/glibc-2.24/csu/../csu/libc-start.c:289

Direct leak of 8 byte(s) in 1 object(s) allocated from:
    #0 0x501ee2 in operator new[](unsigned long) /local/mnt/workspace/bcain_clang_bcain-ubuntu_14427/llvm/utils/release/final/llvm.src/projects/compiler-rt/lib/asan/asan_new_delete.cc:109:3
    #1 0x51e55b in wasm::FuncData::v8_callback(v8::FunctionCallbackInfo<v8::Value> const&) /Users/hisham/wasm-c-api/./src/wasm-v8.cc:1710:18
    #2 0xadb5d0 in v8::internal::FunctionCallbackArguments::Call(v8::internal::CallHandlerInfo*) (/Users/hisham/wasm-c-api/out/example/hello-c+0xadb5d0)
    #3 0xadaae7 in v8::internal::MaybeHandle<v8::internal::Object> v8::internal::(anonymous namespace)::HandleApiCallHelper<false>(v8::internal::Isolate*, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::FunctionTemplateInfo>, v8::internal::Handle<v8::internal::Object>, v8::internal::BuiltinArguments) (/Users/hisham/wasm-c-api/out/example/hello-c+0xadaae7)
    #4 0xada274 in v8::internal::Builtin_Impl_HandleApiCall(v8::internal::BuiltinArguments, v8::internal::Isolate*) (/Users/hisham/wasm-c-api/out/example/hello-c+0xada274)
    #5 0xa3828d  (/Users/hisham/wasm-c-api/out/example/hello-c+0xa3828d)
    #6 0x2d68ef3d54a8  (<unknown module>)
    #7 0x3d1a6bd864f9  (<unknown module>)
    #8 0x361f32184060  (<unknown module>)
    #9 0x9a7842  (/Users/hisham/wasm-c-api/out/example/hello-c+0x9a7842)
    #10 0x361f3210623d  (<unknown module>)
    #11 0x6a4319 in v8::internal::(anonymous namespace)::Invoke(v8::internal::Isolate*, bool, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, int, v8::internal::Handle<v8::internal::Object>*, v8::internal::Handle<v8::internal::Object>, v8::internal::Execution::MessageHandling, v8::internal::Execution::Target) (/Users/hisham/wasm-c-api/out/example/hello-c+0x6a4319)
    #12 0x6a4039 in v8::internal::Execution::Call(v8::internal::Isolate*, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, int, v8::internal::Handle<v8::internal::Object>*) (/Users/hisham/wasm-c-api/out/example/hello-c+0x6a4039)
    #13 0x55da5a in v8::Function::Call(v8::Local<v8::Context>, v8::Local<v8::Value>, int, v8::Local<v8::Value>*) (/Users/hisham/wasm-c-api/out/example/hello-c+0x55da5a)
    #14 0x51d858 in wasm::Func::call(wasm::Val const*, wasm::Val*) const /Users/hisham/wasm-c-api/./src/wasm-v8.cc:1669:33
    #15 0x531aaa in wasm_func_call /Users/hisham/wasm-c-api/src/wasm-c.cc:802:24
    #16 0x504fad in main /Users/hisham/wasm-c-api/example/hello.c:92:7
    #17 0x7f248b94e290 in __libc_start_main /sources/glibc-2.24/csu/../csu/libc-start.c:289

Direct leak of 8 byte(s) in 1 object(s) allocated from:
    #0 0x501ee2 in operator new[](unsigned long) /local/mnt/workspace/bcain_clang_bcain-ubuntu_14427/llvm/utils/release/final/llvm.src/projects/compiler-rt/lib/asan/asan_new_delete.cc:109:3
    #1 0x51e193 in wasm::FuncData::v8_callback(v8::FunctionCallbackInfo<v8::Value> const&) /Users/hisham/wasm-c-api/./src/wasm-v8.cc:1709:15
    #2 0xadb5d0 in v8::internal::FunctionCallbackArguments::Call(v8::internal::CallHandlerInfo*) (/Users/hisham/wasm-c-api/out/example/hello-c+0xadb5d0)
    #3 0xadaae7 in v8::internal::MaybeHandle<v8::internal::Object> v8::internal::(anonymous namespace)::HandleApiCallHelper<false>(v8::internal::Isolate*, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::FunctionTemplateInfo>, v8::internal::Handle<v8::internal::Object>, v8::internal::BuiltinArguments) (/Users/hisham/wasm-c-api/out/example/hello-c+0xadaae7)
    #4 0xada274 in v8::internal::Builtin_Impl_HandleApiCall(v8::internal::BuiltinArguments, v8::internal::Isolate*) (/Users/hisham/wasm-c-api/out/example/hello-c+0xada274)
    #5 0xa3828d  (/Users/hisham/wasm-c-api/out/example/hello-c+0xa3828d)
    #6 0x2d68ef3d54a8  (<unknown module>)
    #7 0x3d1a6bd864f9  (<unknown module>)
    #8 0x361f32184060  (<unknown module>)
    #9 0x9a7842  (/Users/hisham/wasm-c-api/out/example/hello-c+0x9a7842)
    #10 0x361f3210623d  (<unknown module>)
    #11 0x6a4319 in v8::internal::(anonymous namespace)::Invoke(v8::internal::Isolate*, bool, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, int, v8::internal::Handle<v8::internal::Object>*, v8::internal::Handle<v8::internal::Object>, v8::internal::Execution::MessageHandling, v8::internal::Execution::Target) (/Users/hisham/wasm-c-api/out/example/hello-c+0x6a4319)
    #12 0x6a4039 in v8::internal::Execution::Call(v8::internal::Isolate*, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, int, v8::internal::Handle<v8::internal::Object>*) (/Users/hisham/wasm-c-api/out/example/hello-c+0x6a4039)
    #13 0x55da5a in v8::Function::Call(v8::Local<v8::Context>, v8::Local<v8::Value>, int, v8::Local<v8::Value>*) (/Users/hisham/wasm-c-api/out/example/hello-c+0x55da5a)
    #14 0x51d858 in wasm::Func::call(wasm::Val const*, wasm::Val*) const /Users/hisham/wasm-c-api/./src/wasm-v8.cc:1669:33
    #15 0x531aaa in wasm_func_call /Users/hisham/wasm-c-api/src/wasm-c.cc:802:24
    #16 0x504fad in main /Users/hisham/wasm-c-api/example/hello.c:92:7
    #17 0x7f248b94e290 in __libc_start_main /sources/glibc-2.24/csu/../csu/libc-start.c:289

SUMMARY: AddressSanitizer: 824 byte(s) leaked in 3 allocation(s).
make: *** [Makefile:104: run-hello-c] Error 1
```

**Comment by rossberg** [comment]
Ah, I think these ones are different, and I actually see what they are (although I don't understand why they're only detected on some systems). Will fix (but it might take 2 weeks before I get to that).

**Comment by brianchin** [comment]
I believe I fixed these with this PR #61.
